tests/test_specialize.py

```diff
@@ -249,21 +252,40 @@ def f2():
     return f2()(x)
 
 
-@specialize((True, int1, int1))
+@specialize((True, int1, int2))
 def test_poly_with_constants(c, x, y):
     def f1(x, y):
         return x + y
 
     def f2(x, y):
-        return y + x
+        return x * y
 
     def choose(c):
         if c:
             return f1
         else:
             return f2
 
-    return choose(c)(x, 2), choose(not c)(2, y)
+    return choose(c)(x, y), choose(not c)(x, y)
+
+
+@mark.xfail(reason="Distinct contexts are created for 2 and 3, "
+                   "leading to Problem(POLY).")
```

Comment:
Is this hard to fix?

Reply:
Not necessarily, broadly speaking I can see two solutions. One is to have a `max_depth` of 0 for non-boolean types, but that may cut several important optimization opportunities. The other is to make a specialized graph for each possible context, and then checking whether they are all isomorphic, including the inferred types for all nodes. If they are all isomorphic, then we can simply return one of them.

Reply:
If we ever fix this, I think we should use solution 2.

Reply:
I agree.